tests/test_classical_gradients.py

```diff
@@ -231,6 +231,60 @@ def test_linear(self, tol):
         assert np.allclose(auto_grad, correct_grad, atol=tol, rtol=0)
 
 
+class TestGrad:
+    """Unit tests for the gradient function"""
+
+    def test_non_scalar_cost_gradient(self):
+        """Test gradient computation with a non-scalar cost function raises an error"""
+
+        def cost(x):
+            return np.sin(x)
+
+        grad_fn = qml.grad(cost, argnum=[0])
+        arr1 = np.array([0.0, 1.0, 2.0], requires_grad=True)
+
+        with pytest.raises(TypeError, match="only applies to real scalar-output functions"):
+            grad_fn(arr1)
+
+    def test_agrees_with_autograd(self, tol):
+        """Test that the grad function agrees with autograd"""
+
+        def cost(x):
+            return np.sum(np.sin(x) * x[0] ** 3)
+
+        grad_fn = qml.grad(cost)
+        params = np.array([0.5, 1.0, 2.0], requires_grad=True)
+        res = grad_fn(params)
+        expected = autograd.grad(cost)(params)
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_forward_pass_value_storing(self, tol):
+        """Test that the intermediate forward pass value is accessible and correct"""
+
+        def cost(x):
+            return np.sum(np.sin(x) * x[0] ** 3)
+
+        grad_fn = qml.grad(cost)
+        params = np.array([-0.654, 1.0, 2.0], requires_grad=True)
+
+        assert grad_fn.forward is None
+
+        grad = grad_fn(params)
+
+        res = grad_fn.forward
+        expected = cost(params)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # change the parameters
+        params2 = np.array([1.4, 1.0, 2.0], requires_grad=True)
+        grad = grad_fn(params2)
+
+        res = grad_fn.forward
+        expected = cost(params2)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+
```

Comment:
I couldn't see anywhere in the tests a check for `qml.grad` where we have multiple variables, some with `requires_grad=True` and some with `requires_grad=False`, might be worth adding? (I could see this for `qml.jacobian()` though).

Reply:
Have added another test!